File: pack.py
#!/usr/bin/python

import os
import subprocess
import argparse
import json
import copy
# import yaml

# This key is special: it is can not be inherited.
output_path_key = 'output_path'

def get_current_file_path():
    return os.path.dirname(os.path.realpath(__file__))

# Gets the settings json file path for the specified directory.
# @param dir The directory.
def get_settings_path(dir):
    return os.path.join(dir, 'texture_packer.json')

# Merges the current settings with the specified json data.
# @param settings The current settings.
# @param json_data The json data for merging.
def merge_settings(settings, json_data):
    for key in json_data:
        if key == output_path_key:
            continue
        settings[key] = json_data[key]

def pack_with_output_path_array(relative_dir, current_dir, output_dir, output_path_array, settings):
    output_path = os.path.join(output_dir, *output_path_array)
    return pack(relative_dir, current_dir, output_path, settings)

def parse_command(settings):
    command = []

    # Process each key.
    for key in settings:
        if not key.startswith('--'):
            # Not texture packer command line settings.
            continue

        # Multiple variants.
        if key == '--variant':
            for variant in settings[key]:
                command.extend(['--variant', variant])
            continue

        # Normal settings.
        single_options = [
            '--multipack',
            '--enable-rotation',
            '--disable-rotation',
            '--disable-auto-alias',
            '--flip-pvr',
            '--premultiply-alpha',
            '--force-squared',
            '--force-word-aligned',
            '--force-identical-layout',
            '--reduce-border-artifacts',
            '--disable-clean-transparency',
            '--shape-debug'
        ]
        if key in single_options:
            # Settings that do not have value (only key).
            command.append(key)
        else:
            # Must convert to string.
            command.extend([key, str(settings[key])])

    return command

def pack(relative_dir, current_dir, output_path, settings):
    command = parse_command(settings)

    flatten_path = settings.get('flatten_path', True)
    sheet_extension = settings.get('sheet_extension', 'pvr.ccz')
    data_extension = settings.get('data_extension', 'plist')
    input_directories = settings.get('input_directories', ['.'])
    combine_images = settings.get('combine_images', True)

    if not flatten_path:
        command.extend(['--replace', '^=%s/' % relative_dir])

    # Input.
    input_paths = []
    for input_dir in input_directories:
        full_path = os.path.join(current_dir, input_dir)
        for item in os.listdir(full_path):
            path = os.path.join(full_path, item)
            if path.endswith('.png') or path.endswith('.jpg'):
                # PNG or JPG.
                input_paths.append(path)

    commands = []
    if combine_images:
        command.extend(['--sheet', '%s.%s' % (output_path, sheet_extension)])
        command.extend(['--data', '%s.%s' % (output_path, data_extension)])
        command.extend(input_paths)
        commands.append(command)
    else:
        for path in input_paths:
            # http://stackoverflow.com/questions/678236/how-to-get-the-filename-without-the-extension-from-a-path-in-python
            filename = os.path.splitext(os.path.basename(path))[0]

            clone = list(command)
            clone.append(path)
            clone.extend(['--sheet', '%s.%s' % (os.path.join(output_path, filename), sheet_extension)])
            clone.extend(['--data', os.path.join(get_current_file_path(), '.texture_packer_dummy_{v}.plist')])
            commands.append(clone)

    return commands

# Finds all packing commands.
# @param input_dir The input directory.
# @param output_dir The output directory.
def process(input_dir, output_dir):
    # Currently merged settings.
    settings = {}

    # Used for flatten_paths=False
    relative_dir = ''

    # Depth-first search all sub-directories.
    return dfs(relative_dir, input_dir, output_dir, settings)

# DFS the current directory.
# @param relative_dir The relative directory to the input directory.
# @param current_dir The currently processing directory.
# @param output_dit The output directory.
# @param current_settings The currently merged settings.
def dfs(relative_dir, current_dir, output_dir, current_settings):
    # Result.
    commands = []

    # Gets the settings path for the current directory.
    settings_path = get_settings_path(current_dir)

    if os.path.exists(settings_path):
        with open(settings_path) as data_file:
            # http://stackoverflow.com/questions/956867/how-to-get-string-objects-instead-of-unicode-ones-from-json-in-python
            # data = yaml.safe_load(data_file)
            data = json.load(data_file)
            merge_settings(current_settings, data)

            if output_path_key in data:
                # Process command if output_path_key exists.
                output_path_array = data[output_path_key]
                commands.extend(pack_with_output_path_array(relative_dir, current_dir, output_dir, output_path_array, current_settings))

    # Recurse sub-directories.
    for item in os.listdir(current_dir):
        path = os.path.join(current_dir, item)
        if not os.path.isdir(path):
            continue
        child_settings = copy.deepcopy(current_settings)
        child_relative_dir = os.path.join(relative_dir, item)
        child_commands = dfs(child_relative_dir, path, output_dir, child_settings)
        commands.extend(child_commands)

    return commands

def process_commands(input_dir, output_dir):
    current_dir = get_current_file_path()

    commands = process(input_dir, output_dir)

    # Parallelly packing.
    print 'There are ' + str(len(commands)) + ' packing commands.'
    subprocesses = []
    for command in commands:
        args = ['texturepacker']
        args.extend(command)
        subprocesses.append(subprocess.Popen(args))

    # Wait for subprocesses.
    for s in subprocesses:
        s.wait()

def parse_arguments():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '-i',
        '--input',
        nargs = 1,
    )
    parser.add_argument(
        '-o',
        '--output',
        nargs = 1,
    )
    args = parser.parse_args()
    input_dir = args.input[0]
    output_dir = args.output[0]
    return input_dir, output_dir

if __name__ == '__main__':
    input_dir, output_dir = parse_arguments()
    process_commands(input_dir, output_dir)

File: pack.js
/** Pack version 3 */

const assert = require('assert');
const ArgumentParser = require('argparse').ArgumentParser;
const childProcess = require('child_process');
const fetch = require('node-fetch');
const fs = require('fs-extra');
const glob = require('glob');
const JSZip = require('jszip');
const md5 = require('md5');
const pako = require('pako');
const path = require('path');
const util = require('util');

class Cache {
    /** @returns {string} The pack cache path. */
    _getCachePath() {
        return path.join(__dirname, `.pack_cache.json`);
    }

    /**
     * Reads value for the specified key.
     * @param {string} key The desired key.
     * @returns {Promise<string | undefined>} The value for the specified key.
     */
    async _readValue(key) {
        const path = this._getCachePath();
        try {
            const content = await fs.readJson(path);
            return content[key];
        } catch {
            // File not exist or not JSON.
            return undefined;
        }
    }

    /**
     * Sets value for the specified key.
     * @param {string} key The desired key.
     * @param {any} value The value for the desired key.
     */
    async _setValue(key, value) {
        const path = this._getCachePath();
        try {
            const content = await fs.readJson(path);
            content[key] = value;
            await fs.writeJSON(path, content);
        } catch {
            // File not exist or not JSON.
            await fs.writeJSON(path, {
                [key]: value
            });
        }
    }
    /**
     * Checks whether it is necessary to process specified command.
     * @param {any} data
     * @param {string} outputDir
     */
    async check(data, outputDir) {
        const dataHash = md5(JSON.stringify(data));
        const files = await this._readValue(dataHash);
        if (!files) {
            return true;
        }
        for (let i = 0, n = files.length; i < n; ++i) {
            const fileName = files[i].name;
            const fileHash = files[i].hash;
            const filePath = path.join(outputDir, fileName);
            try {
                const content = await fs.readFile(filePath, {
                    encoding: 'base64'
                });
                const currentFileHash = md5(content);
                if (fileHash !== currentFileHash) {
                    return true;
                }
            } catch (ex) {
                // File not found.
                return true;
            }
        }
        return false;
    }

    /** Stores the specified command and its result to the cache. */
    async store(data, outputDir, files) {
        const dataHash = md5(JSON.stringify(data));
        const items = [];
        for (let i = 0, n = files.length; i < n; ++i) {
            const fileName = files[i];
            const filePath = path.join(outputDir, fileName);
            const content = await fs.readFile(filePath, {
                encoding: 'base64'
            });
            const fileHash = md5(content);
            items.push({
                name: fileName,
                hash: fileHash
            });
        }
        await this._setValue(dataHash, items);
    }
}

class CommandProcessor {
    /**
     * Asynchronously process the specified options.
     * @param {any} command
     * @param {string} outputDir
     */
    async process(command, outputDir) {
        throw new Error('Abstract method.');
    }
}

/** Packs using local TexturePacker. */
class LocalProcessor extends CommandProcessor {
    async process(command, outputDir) {
        const params = [];
        params.push('texturepacker');
        params.push(...command.params);
        params.push(...['--sheet', path.join(outputDir, command.sheet)]);
        params.push(...['--data', path.join(outputDir, command.data)]);
        params.push(...command.files);
        await util.promisify(childProcess.exec)(params.join(' '));
    }
}

/** Packs using remote TexturePacker. */
class RemoteProcessor extends CommandProcessor {
    /**
     * @param {string} address 
     */
    constructor(address) {
        super();
        this.address = address;
        this.cache = new Cache();
    }

    async process(command, outputDir) {
        const data = {};
        data.params = command.params;
        data.sheet = command.sheet;
        data.data = command.data;

        const requestId = data.sheet;
        console.log(`${requestId}: start packing`);

        // Basename and Base64-encoded data of files.
        await Promise.all(command.files.map(async item => {
            data.files = data.files || [];
            const content = await fs.readFile(item, {
                encoding: 'base64'
            });
            data.files.push({
                name: path.basename(item),
                data: content,
            });
        }));

        // Ensure the same order between packings.
        data.files.sort((lhs, rhs) => {
            return lhs.name.localeCompare(rhs.name);
        });

        if (!await this.cache.check(data, outputDir)) {
            console.log(`${requestId}: already updated.`);
            return;
        }

        // Send request.
        const stringified = JSON.stringify(data);
        // Use pako with body-parser: https://github.com/expressjs/body-parser/issues/138
        const compressed = pako.gzip(stringified, {
            level: 9,
            memLevel: 9,
            windowBits: 15,
        });
        console.log(`${requestId}: request data size = ${stringified.length} compressed to ${compressed.length}`);
        const response = await fetch(this.address, {
            body: compressed,
            headers: {
                'Content-Encoding': 'gzip',
                'Content-Type': 'application/json',
            },
            method: 'POST',
        }).then(res => res.json());
        console.log(`${requestId}: response data size = ${JSON.stringify(response).length}`);

        if (response.error) {
            console.log(`${requestId}: ${response.error}`);
            return;
        }

        await fs.ensureDir(outputDir);
        const archive = await JSZip.loadAsync(response.result, {
            base64: true,
        });
        const files = Object.keys(archive.files);
        await Promise.all(files.map(async fileName => {
            const filePath = path.join(outputDir, fileName);
            await fs.ensureDir(path.dirname(filePath));
            const file = archive.files[fileName];
            const content = await file.async('base64');
            await fs.writeFile(filePath, content, {
                encoding: 'base64'
            });
        }));
        await this.cache.store(data, outputDir, files);
    }
}

/**
 * DFS the config tree structure.
 * @param node The current node.
 * @param parentOptions The options of the parent of the current node.
 */
function dfsNode(node, parentOptions = {}) {
    // All options of this sub-tree.
    const totalOptions = [];

    // Clone parent's options.
    const options = JSON.parse(JSON.stringify(parentOptions));

    // Apply current node's settings.
    Object.keys(node.options).forEach(key => {
        if (key === 'params') {
            // Append.
            options[key] = options[key] || [];
            options[key].push(...node.options[key]);
        } else {
            // Overwrite.
            options[key] = node.options[key];
        }
    });
    totalOptions.push(options);

    // DFS children.
    if (node.children) {
        node.children.forEach(child => {
            const childOptions = dfsNode(child, options);
            totalOptions.push(...childOptions);
        })
    }

    return totalOptions;
}

/**
 * @param {Array<string | number>} params 
 */
function flattenParams(params) {
    const uniqueParams = {};
    const assign = (key, value) => {
        let values = uniqueParams[key];
        values = values || [];
        if (key === '--variant') {
            // Allow multiple values.
            values.push(value);
        } else {
            if (values.length === 0) {
                values = [value];
            }
        }
        uniqueParams[key] = values;
    }
    let currentValue = undefined;
    // Prefer later parameters.
    for (let i = params.length - 1; i >= 0; --i) {
        const param = params[i];
        if (typeof param === `string` && param.startsWith(`--`)) {
            assert(currentValue !== undefined);
            assign(param, currentValue);
            currentValue = undefined;
        } else {
            currentValue = param;
        }
    }
    const results = [];
    Object.keys(uniqueParams).forEach(key => {
        uniqueParams[key].forEach(item => {
            results.push(...[key, item]);
        })
    });
    return results;
}
/**
 * Parses the specified options and returns an array of commands.
 * @param options The desired options.
 * @param outputDir The desired output directory.
 */
function parseOptions(options, inputDir) {
    const inputFilePatterns = options.input_files;
    const outputPath = options.output_path;
    if (!inputFilePatterns || !outputPath) {
        // No input or output specified.
        return [];
    }
    const params = [];
    if (options.params) {
        options.params.forEach(item => {
            if (item instanceof Array) {
                params.push(...item);
            } else {
                params.push(item);
            }
        });
    }
    const flattenedParams = flattenParams(params);
    if (options.rotation) {
        flattenedParams.push('--enable-rotation');
    } else {
        flattenedParams.push('--disable-rotation');
    }
    if (options.auto_alias) {
        flattenedParams.push('--disable-auto-alias');
    }
    if (options.force_identical_layout) {
        flattenedParams.push('--force-identical-layout');
    }
    const sheetExtension = options.sheet_extension || 'pvr.ccz';
    const dataExtension = options.data_extension || 'plist';

    const inputFiles = [];
    inputFilePatterns.forEach(patternComponents => {
        const pattern = path.join(inputDir, ...patternComponents);
        const files = glob.sync(pattern);
        inputFiles.push(...files);
    });
    return [{
        params: flattenedParams,
        files: inputFiles,
        sheet: `${path.join(...outputPath)}.${sheetExtension}`,
        data: `${path.join(...outputPath)}.${dataExtension}`,
    }];
}

/** Asynchronously process the command. */
async function process(processor, configPath, outputDir) {
    const inputDir = path.dirname(configPath);
    const root = JSON.parse(fs.readFileSync(configPath, 'utf8'));

    // Find all options.
    const totalOptions = dfsNode(root);

    // Parse options.
    const commands = totalOptions.reduce((array, options) => {
        array.push(...parseOptions(options, inputDir));
        return array;
    }, []);

    await Promise.all(commands.map(async command => {
        await processor.process(command, outputDir);
    }));
}

/** Parses arguments from command line. */
function parseArguments() {
    const parser = new ArgumentParser();
    parser.addArgument(['-t', '--type'], {
        choices: ['remote', 'local'],
        required: true,
    });
    parser.addArgument(['-a', '--address'], {
        required: false,
    });
    parser.addArgument(['-i', '--input'], {
        help: 'Path to the settings JSON file',
        nargs: 1,
        required: true,
    });
    parser.addArgument(['-o', '--output'], {
        help: 'Path to the output directory',
        nargs: 1,
        required: true,
    });
    const args = parser.parseArgs();
    console.dir(args);
    return {
        type: args.type,
        address: args.address,
        input: args.input[0],
        output: args.output[0],
    };
}

const {
    type,
    address,
    input,
    output
} = parseArguments();

if (type === 'remote') {
    process(new RemoteProcessor(address), input, output);
} else {
    process(new LocalProcessor(), input, output);
}